Signal/src/ViewControllers/DebugUI/DebugUIMessagesAssetLoader.m

```diff
@@ -0,0 +1,454 @@
+        failure:^(NSURLSessionDataTask *_Nullable task, NSError *requestError) {
+            OWSFail(@"Error downloading url[%@]: %@", fileUrl, requestError);
+            failure();
+        }];
+}
+
+#pragma mark -
+
++ (DebugUIMessagesAssetLoader *)fakePngAssetLoaderWithImageSize:(CGSize)imageSize
+                                                backgroundColor:(UIColor *)backgroundColor
+                                                      textColor:(UIColor *)textColor
+                                                          label:(NSString *)label
+{
+    OWSAssert(imageSize.width > 0);
+    OWSAssert(imageSize.height > 0);
+    OWSAssert(backgroundColor);
+    OWSAssert(textColor);
+    OWSAssert(label.length > 0);
+
+    DebugUIMessagesAssetLoader *instance = [DebugUIMessagesAssetLoader new];
+    instance.mimeType = OWSMimeTypeImagePng;
+    instance.filename = @"image.png";
+    __weak DebugUIMessagesAssetLoader *weakSelf = instance;
+    instance.prepareBlock = ^(ActionSuccessBlock success, ActionFailureBlock failure) {
+        [weakSelf ensurePngAssetLoaded:imageSize
+                       backgroundColor:backgroundColor
+                             textColor:textColor
+                                 label:label
+                               success:success
+                               failure:failure];
+    };
+    return instance;
+}
+
+- (void)ensurePngAssetLoaded:(CGSize)imageSize
+             backgroundColor:(UIColor *)backgroundColor
+                   textColor:(UIColor *)textColor
+                       label:(NSString *)label
+                     success:(ActionSuccessBlock)success
+                     failure:(ActionFailureBlock)failure
+{
+    OWSAssert(success);
+    OWSAssert(failure);
+    OWSAssert(self.filename.length > 0);
+    OWSAssert(self.mimeType.length > 0);
+    OWSAssert(imageSize.width > 0 && imageSize.height > 0);
+    OWSAssert(backgroundColor);
+    OWSAssert(textColor);
+    OWSAssert(label.length > 0);
+
+    if (self.filePath) {
+        success();
+        return;
+    }
+
+    NSString *filePath = [OWSFileSystem temporaryFilePathWithFileExtension:@"png"];
+    UIImage *image =
+        [self createRandomPngWithSize:imageSize backgroundColor:backgroundColor textColor:textColor label:label];
+    NSData *pngData = UIImagePNGRepresentation(image);
+    [pngData writeToFile:filePath atomically:YES];
+    self.filePath = filePath;
+    OWSAssert([NSFileManager.defaultManager fileExistsAtPath:filePath]);
+    success();
+}
+
+- (nullable UIImage *)createRandomPngWithSize:(CGSize)imageSize
+                              backgroundColor:(UIColor *)backgroundColor
+                                    textColor:(UIColor *)textColor
+                                        label:(NSString *)label
+{
+    OWSAssert(imageSize.width > 0 && imageSize.height > 0);
+    OWSAssert(backgroundColor);
+    OWSAssert(textColor);
+    OWSAssert(label.length > 0);
+
+    CGRect frame = CGRectZero;
```

Comment:
When we generate images, we stamp them with some text to ensure that:

* We're cropping them properly (centered).
* We're scaling them properly (not distorted).